### ISSUE TITLE
`map_coordinates` does not handle coordinate arrays with more than two ranks

### ISSUE THREAD

**Issue by beam2d** [cupy/cupy#2345]
(Originally reported at [this Stackoverflow post](https://stackoverflow.com/questions/57261136/cupy-map-coordinates-gives-out-shape-is-mismatched-error))

The Scipy counterpart accepts the `coordinates` array of shape `(input_ndim, output_dims...)` with multiple axes in `output_dims...`, while the current CuPy does not.

The direct cause is maybe [this weight initialization logic](https://github.com/cupy/cupy/blob/master/cupyx/scipy/ndimage/interpolation.py#L124-L127), which does not take into account the multiple output dims.

* Conditions
```
CuPy Version          : 7.0.0b2
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10000
CUDA Driver Version   : 10000
CUDA Runtime Version  : 10000
cuDNN Build Version   : 7500
cuDNN Version         : 7500
NCCL Build Version    : 2402
NCCL Runtime Version  : 2402
```

* Code to reproduce
```py3
import cupy
import cupyx.scipy.ndimage
x = cupy.random.rand(3, 3)
d = cupy.random.randint(3, size=(2, 4, 5))
cupyx.scipy.ndimage.map_coordinates(x, d)
```

* Error messages, stack traces, or logs
```
Traceback (most recent call last):
  File "foo.py", line 6, in <module>
    cupyx.scipy.ndimage.map_coordinates(x, d)
  File "/home/beam2d/miniconda3/envs/py37/lib/python3.7/site-packages/cupyx/scipy/ndimage/interpolation.py", line 134, in map_coordinates
    weight *= coordinates_ceil[i] - coordinates[i]
  File "cupy/core/core.pyx", line 1047, in cupy.core.core.ndarray.__imul__
  File "cupy/core/_kernel.pyx", line 826, in cupy.core._kernel.ufunc.__call__
  File "cupy/core/_routines_manipulation.pyx", line 69, in cupy.core._routines_manipulation._broadcast_core
ValueError: operands could not be broadcast together with shapes (4,), (4, 5), (4,)
```

**Comment by tvercaut** [comment]
I faced the same issue. My current workaround for cupy is to flatten and then reshape as illustrated in this (probably naive) snippet:
```python
%matplotlib inline
import numpy as np
from PIL import Image
import timeit
import requests
url='https://upload.wikimedia.org/wikipedia/commons/b/b7/Normal_axial_T2-weighted_MR_image_of_the_brain.jpg'
im = Image.open(requests.get(url, stream=True).raw)

np_im = np.array(im)
grid_x, grid_y = np.meshgrid(np.arange(np_im.shape[1]), np.arange(np_im.shape[0]))
dx = 2*np.random.randn(*np_im.shape[:-1])
dy = 2*np.random.randn(*np_im.shape[:-1])
def_grid_x = grid_x + dx
def_grid_y = grid_y + dy

from scipy import ndimage
print('scipy')
%timeit \
out_im_r = ndimage.map_coordinates(np_im[:,:,0], [def_grid_y, def_grid_x], order=1); \
out_im_g = ndimage.map_coordinates(np_im[:,:,1], [def_grid_y, def_grid_x], order=1); \
out_im_b = ndimage.map_coordinates(np_im[:,:,2], [def_grid_y, def_grid_x], order=1); \
out_im = np.stack((out_im_r, out_im_g, out_im_b),2)

import cv2 as cv
print('\nopencv cpu')
%timeit out_im = cv.remap(np_im, np.float32(def_grid_x), np.float32(def_grid_y), cv.INTER_LINEAR)

import cupy as cnp
from cupyx.scipy import ndimage as cndimage
cnp_im = cnp.asarray(np_im)
cdef_grid_x = cnp.asarray(def_grid_x.flatten())
cdef_grid_y = cnp.asarray(def_grid_y.flatten())
cdef_grid = cnp.stack((cdef_grid_y, cdef_grid_x),0)
print('\ncupy on device')
%timeit \
out_im_r = cndimage.map_coordinates(cnp_im[:,:,0], cdef_grid, order=1).reshape(cnp_im.shape[:-1]); \
out_im_g = cndimage.map_coordinates(cnp_im[:,:,1], cdef_grid, order=1).reshape(cnp_im.shape[:-1]); \
out_im_b = cndimage.map_coordinates(cnp_im[:,:,2], cdef_grid, order=1).reshape(cnp_im.shape[:-1]); \
out_im = cnp.stack((out_im_r, out_im_g, out_im_b),2)

import torch
import torch.nn.functional as nnf
print('\ntorch with ' + torch.cuda.get_device_name(0))
cuda = torch.device('cuda')
torch_im = torch.from_numpy(np.float64(np_im)).unsqueeze(0).permute(0,3,1,2).cuda()
gx = 2.0*def_grid_x/(np_im.shape[1]-1)-1.
gy = 2.0*def_grid_y/(np_im.shape[0]-1)-1.
torch_grid = torch.from_numpy(np.stack((gx,gy),2)).unsqueeze(0).cuda()
%timeit torch_out = nnf.grid_sample(torch_im, torch_grid, mode='bilinear', align_corners=False, padding_mode='reflection').squeeze().permute(1,2,0)
```

Note that as per #2844, I was interested in faster image resampling but it turns out the performance I achieve is limited in comparison to opencv (cpu) and pytorch as illustrated in the following figures from running it in a GPU session on google colab:
```
scipy
10 loops, best of 3: 24.8 ms per loop

opencv cpu
1000 loops, best of 3: 1.28 ms per loop

cupy on device
100 loops, best of 3: 5.22 ms per loop

torch with Tesla T4
10000 loops, best of 3: 148 µs per loop
```

**Comment by grlee77** [comment]
I think this issue can be closed (the bug was fixed by #2813).

The separate issue of poor performance reported by @tvercaut should be resolved by #3166.
